mscore/pagesettings.cpp

```diff
@@ -270,311 +399,428 @@ void PageSettings::orientationClicked()
-      if (mmUnit)
-            val /= INCH;
-      preview->score()->style().set(Sid::pageEvenBottomMargin, val);
-      updatePreview(1);
+      preview->score()->setPageNumberOffset(val-1);
+      updatePreview();
       }
 
 //---------------------------------------------------------
-//   spatiumChanged
+//   unitsChanged
 //---------------------------------------------------------
 
-void PageSettings::spatiumChanged(double val)
+void PageSettings::unitsChanged()
       {
-      val *= mmUnit ? DPMM : DPI;
-      double oldVal = preview->score()->spatium();
-      preview->score()->setSpatium(val);
-      preview->score()->spatiumChanged(oldVal, val);
-      updatePreview(0);
+      QPageLayout::Unit u = QPageLayout::Unit(unitsList->currentIndex());
+      MStyle& style = preview->score()->style();
+      style.pageOdd().setUnits(u);
+      style.pageEven().setUnits(u);
+      updateWidgets(true);
       }
 
+///////////////////// Buttons /////////////////////////////
 //---------------------------------------------------------
-//   pageOffsetChanged
+//   setToDefault
 //---------------------------------------------------------
 
-void PageSettings::pageOffsetChanged(int val)
+void PageSettings::setToDefault()
       {
-      preview->score()->setPageNumberOffset(val-1);
-      updatePreview(0);
+      MStyle& style = preview->score()->style();
+      MStyle& def   = MScore::defaultStyle();
+      style.setPageOdd(def.pageOdd());
+      style.setPageEven(def.pageEven());
+      style.setPageSize(def.pageSize());
+      updateWidgets();
+      updatePreview();
       }
 
 //---------------------------------------------------------
-//   pageHeightChanged
+//   applyToAllParts
 //---------------------------------------------------------
 
-void PageSettings::pageHeightChanged(double val)
+void PageSettings::applyToAllParts()
       {
-      double val2 = pageWidth->value();
-      if (mmUnit) {
-            val /= INCH;
-            val2 /= INCH;
-            }
-      pageGroup->setCurrentIndex(0);      // custom
-      preview->score()->style().set(Sid::pageHeight, val);
-      preview->score()->style().set(Sid::pageWidth, val2);
-
-      updatePreview(1);
+      cs->startCmd();
+      for (Excerpt* e : cs->excerpts())
+            applyToScore(e->partScore(), false);
+      cs->endCmd();
       }
 
 //---------------------------------------------------------
-//   pageWidthChanged
+//   applyToScore
 //---------------------------------------------------------
 
-void PageSettings::pageWidthChanged(double val)
+void PageSettings::applyToScore(Score* score, bool runCmd)
       {
-      setMarginsMax(val);
+      Score* prev = preview->score();
+      QPageSize& psize = prev->style().pageSize();
+      MPageLayout& odd = prev->style().pageOdd();
+      MPageLayout& even = prev->style().pageEven();
+
+      // Has anything changed? This prevents extra undo commands on the stack
+      // when the user clicks Apply or OK with no changes.
```

Comment:
Actually here is the chance to fix the whole issue with rounding values on measurement units changes, not only extra undo stack entries. There is no need to change the actual value until the user explicitly changes it, and it should not depend on which units are selected. And only the changed values should be actually overwritten, you should not just overwrite all values in case one small setting is changed. This seems to be a simple and logical rule, and conforming it would actually fix the most part of the issues you are trying to fix. I tried to propose it in the forum topic (for example, [here](https://musescore.org/en/node/287111#comment-910006)), but you seem to use the approach with largely changing the way MuseScore reads the style values from the file — is that really necessary to do?

Reply:
Changing the dialog to operate the way editstyle.ui does, with no Apply button, is possible, but it's a big change.  And I don't know that it would help with anything.  I tried to fix all of this with an upgrade to the way MuseScore handles page settings.   I could have hacked something, but I didn't want to do that.

This belongs in a more general spot, not a review comment, but here is the basic idea:
MPageLayout stores the values in the user units.  That way the source values are exactly the way the user wants them.
MPageLayout then provides ways to extract unrounded, converted values for MuseScore, in whatever units MuseScore needs.

Reply:
> with no Apply button

I don't propose not having Apply button, I propose just changing the way both Apply and OK buttons work. The proposed changes should probably be able to avoid most of cases of unintended page settings changes which was, as far as I understand, one of the most important issues originally.

Concerning MPageLayout, it may indeed be convenient to have such a class to be a "bridge" between the human-readable values and MuseScore internal values. In the approach I propose though it should be not necessary to have this class built directly into `MStyle`, and the changes could limit just to reworking the page settings dialog itself.

Reply:
I'll think about this, but I don't know how it would work off the top of my head.  The page settings end up as styles when reading/writing the file.  The conversion to/from styles would have to be elsewhere.
How would this work when opening an existing score from a file?  Score is in libmscore.  MStyle is in libmscore.  What part of opening an existing score is in mscore?

Reply:
`MStyle::pageOdd()` is used by these libmscore modules, in addition to style.cpp, read206.cpp, read301.cpp:
layout.cpp: `checkDivider()` and `Score::checkSystem()`
page.cpp: all the margins functions, `tm()` `lm()` `bm()` and `rm()`
score.cpp: `Score::Score()` and `innerWidth()` + `innerHeight()`
undo.cpp: my new `ChangePageSettings` class

This is `MPageLayout` providing unrounded values in MuseScore units.  I don't see a way to encapsulate this in the mscore project.  It must be available in libmscore.  And it's not a static instance of anything in class `MScore`.  These are individual object that pertain to a score, not to the application.

If undo.cpp is in libmscore, and I need to undo this stuff, how can it be only in mscore?  If there's a way to do this that I'm not understanding, please tell me.

Reply:
One solution that comes to mind is to create a static MScore variable:
`std::map<Score*, PageSettings&> _pageSettings;`

It would require a new `PageSettings` class in mscore.  Then the libmscore code that needs the page settings values could look up the score's objects and values in the map.

Reply:
btw - Marc is having a similar issue now with the `_chordsList` member variable of `MStyle`.  It is the one other style value not in `_values`, and it is also an object, not a Variant.  I'm unsure of the solution he is pursuing.
https://t.me/musescoreeditorchat/34248

`_chordsList` is different than the page settings objects because it does not duplicate any data in `_values`.  I'm not sure how significant a difference that is within this discussion.

If `MStyle` is best as purely a container for `_values` and `_precomputedValues`, then all the object member variables should move elsewhere.  Otherwise they must be managed outside of the rest of the styles that are so conveniently arranged in the `_values` array.

**Here's another possibility:**  QVariant can contain objects too, right?
How about adding new Sids for pageSizeObj, pageOddObj, pageEvenObj?
There would be special cases to avoid writing them to the file.
Would this be beneficial? Hmmm.
It would eliminate the need for calling `fromPageLayout301()` from `Score::clone()`.  It would also require additional code in `XMLReader`.